### ISSUE TITLE
Hairpinning config with `pf`?

### ISSUE THREAD

**Issue by za-creature** [miniupnp/miniupnp#793]
Sorry to resurrect #257, but I don't believe that the rational used to close it can be applied universally.

On OpenBSD at least, pf's capabilities to perform packet forwarding are (intentionally?) severely limited. My understanding is that any one packet can only ever be matched at most twice: first when coming `in` and the second time when being routed `out`.

This causes problems with hairpinning, as described in [the documentation](https://man.openbsd.org/pf.conf#rdr-to) as well as [the faq](https://www.openbsd.org/faq/pf/rdr.html#rdrnat) in that the network stack does not "simulate" packet forwarding:

A packet from `192.168.0.42:1234` to `203.0.113.99:4321` (WAN) is matched by a `on $lan` rule but not an `on egress` rule, instead the kernel recognizes the WAN IP as itself and stops forwarding; even if it were possible to use `on $lan to egress rdr-to egress` (which for me didn't work), this would probably trip antispoof filters.

Seems like the recommended way to do hairpinning is via a `rdr-to` applied inbound on **both** egress and lan + a `nat-to` applied outbound on lan when the input was lan (`received-on` in this case appears to be some authoritary kernel magic):

```
pass in quick on {$lan egress} inet proto {tcp, udp} to egress port 4321 rdr-to 192.168.0.42 port 1234
pass out on $lan inet to 192.168.0.42 port 1234 received-on $lan nat-to $lan
```

While this overzealous, generic rule can handle the NAT aspect of hairpinning
```
pass out on $lan inet to $lan:network received-on $lan nat-to $lan
```
it's still useless on its own without also including $lan into the list of interfaces that the inbound rules apply to.

---

In conclusion, unless I'm missing something, I agree with [this statement](https://github.com/miniupnp/miniupnp/issues/257#issuecomment-1042616575) and propose adding the above recommended rule-pair for every opened port, presumably under an extra flag "pf_allow_hairpinning"

**Comment by miniupnp** [comment]
Currently, if asked for a UDP port mapping  54321 => 192.168.0.42:12345
the following pf rule is created : 
```
   pass in quick on $ext_if inet proto udp from any to any port 54321 \
             label "test label" rdr-to 192.168.0.42 port 12345
```
additionally if symetric NAT / NAT Port Triggering is enabled, the following is also created for UDP (not for TCP) : 
```
 nat on $ext_if inet proto udp from 192.168.0.42 port 12345 to any -> (re0) port 54321
```

Should theses rules be modifier ?
What rules should be added ?

**Comment by za-creature** [comment]
TLDR:

```
-pass in quick on $ext_if
+pass in quick on {$int_if, $ext_if}
```
If there's more than one int_if configured, include all of them.

While you're there, you could also restrict destinations to just the interface's external ip, though this is optional:
```
-from any to any
+from any to $ext_if
```

This is the minimum necessary to allow hairpinning on BSD, but insufficient by itself. You also need a backwards-nat rule for matched packets:

```
pass out on {$int_if1, $int_if2, ...} \
    inet proto {tcp, udp} \
    to 192.168.0.42 port 12345 \
    received-on {$int_if1, $int_if2, ...} \
    nat-to $ext_if
```

`received-on` is not that well documented, but I believe it's set by the kernel when forwarding packets so you can think of it as a "if it's spoofed, then it's spoofed by the good guys(tm)"

The reason this second rule is not strictly necessary is because it can be (albeit unwisely) matched generically with a custom external rule (feel free to use this when testing):

```
-to 192.168.0.42 port 12345
+to {$int_if1:network, $int_if2:network, ...}
```

---

The reason for all of the above is that nat hairpinning is surprisingly expensive (and quite port-heavy) which is likely why the openbsd devs didn't bother to add an easy to use "enable" button, and probably also why they went out of their way to encourage alternative solutions (like split horizon dns) in the official docs.

That being said, hairpinning is required by the [**good** nat rfc](https://datatracker.ietf.org/doc/html/rfc4787#section-6) so it's worth implementing here too.

Basically, it goes something like this:

If Alice (192.168.0.100) wants to talk to Bob (192.168.0.200) via bob's externally forwarded port, it attempts to send a message to 1.2.3.4 which is forwarded to the router since it's outside her subnet.

The router sees`*` this packet destined for itself, determines that it should forward it to Bob and redirects it. While Bob could respond to Alice directly since they're on the same subnet, her firewall would drop the response since she was trying to talk to 1.2.3.4 not 192.168.0.200. Thus Bob needs to reply via the router, which is where the second nat rule comes into play.

Official docs recommend nat-ing these redirected packets to $int_if (presumably because that saves an external port) but the RFC instead recommends nat-ing to $ext_if instead. I don't know that it matters all that much either way, though I guess I'd go with $ext_if since it makes it very clear that all traffic goes through the router (if I look at 192.168 I immediately think "oh, my 10G switch handles this traffic so it's fine")

`*` currently it doesn't because it only expects this packet on $ext_ip but it's received on $int_if

---

Regarding the final `nat on` rule, I think that's only supported by freebsd's fork of `pf`. Having never used port triggering myself, I'll go with a firm "no comment" on this one.

**Comment by miniupnp** [comment]
So me must change
```
  pass in quick on $ext_if inet proto udp from any to any port 54321 \
             label "test label" rdr-to 192.168.0.42 port 12345
```
to
```
  pass in quick on {$int_if, $ext_if} inet proto udp from any to $extif port 54321 \
             label "test label" rdr-to 192.168.0.42 port 12345
```
and add the other "pass out" rule.

That is complicated because I believe a  `pass in quick on {$int_if, $ext_if} [...]` rule into pf.conf will translate into 2 rules in pf : `pass in quick on $int_if [...]`  and ` pass in quick on $ext_if [...]`

**Comment by za-creature** [comment]
Well three to be exact because this third reverse-nat rule...
```
pass out quick on $int_if inet proto udp to 192.168.0.42 port 12345 received-on $int_if nat-to $ext_if
```
...is also _strongly_ recommended since having it:
1. does not require external rules
2. is probably more secure since it's limited in scope and temporary (e.g. only while the port is being forwarded)

But six in practice since a host can request both tcp and udp forwarding so there should be a `$proto` var in there somewhere.

I'm happy to write up a pull request if you'd like me to, but heads up: i haven't written production C code for more than 10 years
